template: When rendering exception from Jinja, include exception name.

pfalcon opened the pull request:
This will give basic context or _what_ the error is, otherwise there're
messages like:

fatal: [tcwg.ci.linaro.org] => {'msg': 'files/private/jenkins_ssh_publishers.xml.tcwg-ci', 'failed': True}

Which give a filename related to error, but not an idea what error is.

Addresses #7103.

pfalcon committed:
template: When rendering exception from Jinja, include exception name.

This will give basic context or *what* the error is, otherwise there're
messages like:

fatal: [tcwg.ci.linaro.org] => {'msg': 'files/private/jenkins_ssh_publishers.xml.tcwg-ci', 'failed': True}

Which give a filename related to error, but not an idea what error is.

Addresses #7103.

mpdehaan commented:
I think a better solution might be to detect and catch this type of error message and then return a different string message that explains the error in greater depth.   If possible.

pfalcon commented:
Yes, makes sense for "template not found" error, but the problem is that who knows how many more exceptions Jinja2 can throw, and for what reasons. All of them would be hard to predict... What about idea in #7103 - to provide CLI switch to show exception tracebacks to user?

Otherwise, should I give this another try or leave it to you guys? My guess that such check would belong to lib/ansible/utils/template.py , which appears to be proxy between Jinja2 and rest of Ansible, is that good place or do I miss something?

pfalcon committed:
template: Handle TemplateNotFound to render more helpful error message.

At the point the exception is handled, it is likely due to error loading
a sub-template included from main template. Besides file not found, it
can be caused also by include path failing Jinja2 checks.

pfalcon commented:
Ok, I added user-friendly handling of TemplateNotFound, while still left previous generic handling of jinja2 exceptions in case something else might pop up in there.

Example of reporting:

```
fatal: [tcwg.ci.linaro.org] => {'msg': 'AnsibleError: file: /home/pfalcon/devel/linaro/linaro-jenkins-tools/ansible/roles/jenkins-ssh-publish/templates/jenkins.plugins.publish_over_ssh.BapSshPublisherPlugin.xml, error: Cannot find/not allowed to load (include) template ../../ansible-secrets-jenkins/jenkins_ssh_publishers.xml.tcwg-ci', 'failed': True}
```

Let me know if this needs further tweaks.

pfalcon commented:
ping

jimi-c commented:
I merged this in while squashing the commits, so it was merged despite github's UI indicating it was simply closed. Thanks!
